crates/subspace-networking/src/utils/rate_limiter.rs

```diff
@@ -34,13 +34,35 @@ pub(crate) const REGULAR_CONCURRENT_TASKS_BOOST_PER_PEER: usize = 25;
 /// Defines the minimum size of the "connection limit semaphore".
 const MINIMUM_CONNECTIONS_SEMAPHORE_SIZE: usize = 3;
 
+/// A hint for [`RateLimiter`]. It indicates whether an operation follows the Kademlia requests and
+/// assumes that the connection to the peer already exists. It will prevent the [`RateLimiter`] to
+/// obtain a semaphore permit from the connection related semaphore.
+/// It mostly relates to `get-providers` Kademlia request and not for `get-closest-peers` because
+/// in our configuration `get-providers` is guaranteed to return providers from the exact peer and
+/// a connection to the peer will last at least `connection timeout` seconds.
```

Comment:
I get that we have a situation when we have one request followed by the other and connection is already established in that case, but why should the caller care about this or why would it be an issue to obtain a permit in this case? Node runner intentionally increases this semaphore specifically of that reason.

Reply:
We don't increase semaphores for connections only for substreams because the value depends on the connection limits that are constant during the application run. We reused `ResizableSemaphore` type but didn't use its changing capabilities. 

If we treat requests independently we frequently run into a situation when we start multiple Kademlia requests and after that, we wait for other independent operations, existing connections close, and we need to establish them again. The hints here allow to relax blocking in the most used case and repurpose opened connections. It works without hints at all but slower.

Reply:
> We don't increase semaphores for connections only for substreams because the value depends on the connection limits that are constant during the application run. We reused ResizableSemaphore type but didn't use its changing capabilities.

I see, should have used regular tokio semaphore then.

> If we treat requests independently we frequently run into a situation when we start multiple Kademlia requests and after that, we wait for other independent operations, existing connections close, and we need to establish them again. The hints here allow to relax blocking in the most used case and repurpose opened connections. It works without hints at all but slower.

I don't see how it protects us from running out of connection limit then, which was the purpose of having that semaphore.

Reply:
> I see, should have used regular tokio semaphore then.

I used it originally but it's implementation leaks the lifetime all up in the call hierarchy which is very inconvenient.

> I don't see how it protects us from running out of connection limit then, which was the purpose of having that semaphore.

That is correct and sometimes we get failed retrievals, however, the probability on my tests was less than 0.1%. The current implementation limits simultaneous Kademlia quieries and doesn't regulate subsequent request-response. It's still an improvement. An alternative would be to introduce `--strict-request-separation` flag that will enable full separation again for users experiencing a lot of failed requets on their configuration.

Reply:
> I used it originally but it's implementation leaks the lifetime all up in the call hierarchy which is very inconvenient.

There is an owned permit there is you use `Arc<Semaphore>`.

> That is correct and sometimes we get failed retrievals, however, the probability on my tests was less than 0.1%. The current implementation limits simultaneous Kademlia quieries and doesn't regulate subsequent request-response. It's still an improvement. An alternative would be to introduce --strict-request-separation flag that will enable full separation again for users experiencing a lot of failed requets on their configuration.

I feel something is wrong here if we introduce limits just to bypass them in typical case, it makes little sense logically even if happens to work in practice most of the time.

If you want to specifically be able to make request after Kademlia lookup then we need to introduce additional abstraction that will fist acquire all necessary permits, then give you a handle that can be used to make several requests in a series without giving permits away.

Something like this:
```rust
struct NodeRequestsBatchHandle {}

impl NodeRequestsBatchHandle {
    pub async fn get_closest_peers(&mut self, ...) -> ...;
    pub async fn send_generic_request(&mut self, ...) -> ...;
    ...
}

impl Node {
    pub fn async get_requests_batch_handle(&self) -> NodeRequestsBatchHandle;
}
```

Note that `NodeRequestsBatchHandle` takes `&mut self`, so it doesn't allow to use single permit collection for multiple concurrent requests and violate imposed requirements that way.

Reply:
I changed the semantics and removed hints for rate limiter.